src/main/java/com/aws/greengrass/status/FleetStatusService.java

```diff
@@ -295,24 +296,40 @@ public void schedulePeriodicFleetStatusDataUpdate(boolean isDuringConnectionResu
                 initialDelay, periodicPublishIntervalSec, TimeUnit.SECONDS);
     }
 
+    private AtomicReference<OverallStatus> getOverAllStatus() {
+        Instant now = Instant.now();
+        AtomicReference<OverallStatus> overAllStatus = new AtomicReference<>(OverallStatus.HEALTHY);
+        this.kernel.orderedDependencies().forEach(service -> {
+            serviceFssTracksMap.put(service, now);
+            overAllStatus.set(getOverallStatusBasedOnServiceState(overAllStatus.get(), service));
+        });
+        return overAllStatus;
+    }
+
     @SuppressWarnings("PMD.UnusedFormalParameter")
     private void handleServiceStateChange(GreengrassService greengrassService, State oldState,
                                           State newState) {
+
+        Set<GreengrassService> erroredComponentSet = new HashSet<>();
+
         // Not reporting status of other components in errored component message because a deployment could
         // be in-progress and other component's fleet config arn may not have updated
         if (newState.equals(State.ERRORED)) {
-            Set<GreengrassService> erroredComponentSet = new HashSet<>();
             erroredComponentSet.add(greengrassService);
+            if(!isDeploymentInProgress.get()){
+                //Do not update the recovery state later for ongoing deployment
+                erroredComponentsSet.add(greengrassService);
+            }
 
             // Evaluate overall status based on current state of all components
-            Instant now = Instant.now();
-            AtomicReference<OverallStatus> overAllStatus = new AtomicReference<>(OverallStatus.HEALTHY);
-            this.kernel.orderedDependencies().forEach(service -> {
-                serviceFssTracksMap.put(service, now);
-                overAllStatus.set(getOverallStatusBasedOnServiceState(overAllStatus.get(), service));
-            });
+            uploadFleetStatusServiceData(erroredComponentSet, getOverAllStatus().get(), null, Trigger.COMPONENT_STATUS_CHANGE);
+            return;
+        }
 
-            uploadFleetStatusServiceData(erroredComponentSet, overAllStatus.get(), null, Trigger.ERRORED_COMPONENT);
+        // if the reported errored component is recovered, remove the component from the set and update the component status
+        if (!isDeploymentInProgress.get() && greengrassService.reachedDesiredState() && erroredComponentsSet.remove(greengrassService)) {
+            erroredComponentSet.add(greengrassService);
+            uploadFleetStatusServiceData(erroredComponentSet, getOverAllStatus().get(), null, Trigger.COMPONENT_STATUS_CHANGE);
```

Comment:
So, lets say that you have 2 services which recover from an errored state outside a deployment. FSS state change handler will receive 2 updates.
1. Update 1 adds ServiceA to `erroredComponentState` and triggers upload to FSS. ServiceA recovery is published.
2. Update 2 adds ServiceB to `erroredComponentState` and triggers upload to FSS again. This time both ServiceA and ServiceB recovery is published.

Is this the desired design?

Reply:
Yes, we only update partial fss message to that specific component, the design logic is pretty much the same as errored component, we update whichever failed. The only difference is we only update the recovery ones when they happen outside of deployments